R/gh_token.R

```diff
@@ -164,21 +180,162 @@ gh_auth <- function(token) {
+  host <- sub("/.*$", "", rest)              #         github.uni.edu
+  paste0(prot, host)                         # https://github.uni.edu
+}
+
+normalize_host <- function(x) {
+  sub("api.github.com", "github.com", x)
+}
+
+get_hosturl <- function(url) {
+  url <- get_baseurl(url)
+  normalize_host(url)
+}
+
+get_apiurl <- function(url) {
+  host_url <- get_hosturl(url)
+  prot_host <- strsplit(host_url, "://", fixed = TRUE)[[1]]
+  if (is_github_dot_com(host_url)) {
+    paste0(prot_host[[1]], "://api.github.com")
+  } else {
+    paste0(host_url, "/api/v3")
+  }
+}
+
+is_github_dot_com <- function(url) {
+  url <- get_baseurl(url)
+  url <- normalize_host(url)
+  grepl("^https?://github.com", url)
+}
+
+# inlined from credentials and generalized to non-github.com URL
+# is_interactive() is an unexported helper defined in utils.R
+# morally, is rlang::is_interactive()
+set_github_pat2 <- function(api_url = default_api_url(),
+                            force_new = FALSE,
+                            validate = is_interactive(),
+                            verbose = validate) {
+  if (!can_load("credentials")) {
+    return(FALSE)
+  }
+
+  host_url <- get_hosturl(api_url)
+  prot_host <- strsplit(host_url, "://", fixed = TRUE)[[1]]
+  if (is_github_dot_com(api_url)) {
+    # what's currently in credentials
+    # pat_user <- Sys.getenv("GITHUB_PAT_USER", 'PersonalAccessToken')
+    # TODO: I propose we change the user env var to GITHUB_USER
+    user_env_var <- "GITHUB_PAT_USER"
+    token_env_var <- "GITHUB_PAT"
+  } else {
+    user_env_var <- paste0("GITHUB_USER_", slugify_url(api_url))
+    token_env_var <- paste0("GITHUB_PAT_", slugify_url(api_url))
+  }
+  api_url <- get_apiurl(api_url)
+
+  # The username doesn't have to be real, Github seems to ignore username for PATs
+  pat_user <- Sys.getenv(user_env_var, "PersonalAccessToken")
+  # credentials does: pat_url <- sprintf("https://%s@github.com", pat_user)
+
+  pat_url <- sprintf(
+    "%s://%s@%s",
+    prot_host[[1]], pat_user, prot_host[[2]]
+  )
+
+  if (isTRUE(force_new)) {
+    credentials::git_credential_forget(pat_url)
+  }
+
+  # credentials does the "hack" unconditionally, but I think this is better
+  if (is_interactive()) {
+    if(isTRUE(verbose)) {
+      message2("If prompted for GitHub credentials, enter your PAT in the password field")
+    }
+    askpass <- Sys.getenv('GIT_ASKPASS')
+    if (nchar(askpass)) {
+      # Hack to override prompt sentence to say "Token" instead of "Password"
+      withr::local_envvar(c(
+        GIT_ASKTOKEN = askpass,
+        GIT_ASKPASS = system.file(
+          "ask_token.sh", package = "gh", mustWork = TRUE
+        ),
+        GIT_ASKTOKEN_NAME = sprintf(
+          "Personal Access Token (PAT) for %s", host_url
+        )
+      ))
+    }
+  } else {
+    withr::local_envvar(c(
+      GIT_ASKPASS = system.file(
+        "empty_token.sh", package = "gh", mustWork = TRUE
```

Comment:
This is the only, clean way I have found to reliably prevent an attempt to interact with the user in semi-interactive settings, i.e. when testing or checking.

Reply:
I don't think this'll work on Windows. But we can think about a different solution later.

Reply:
Maybe @jeroen can tell us. A very similar method is used in the other branch (interactive session), so I don't see how that could work and this could not?

Reply:
Neither work. You can't rely on `sh` on Windows.

Reply:
The git for windows credential for Windows has its own GUI. It completely ignores these environment variables.

Reply:
Also, note that Git for Windows brings its own bash shell environment and everything, because git depends on this. I don't think it's possible to have a working git installation, but not sh.

Reply:
Sure, but `git` is on the path, and `sh` is not.

Reply:
But let's merge it and we'll see if it works. :)